File: django_backend/simple_grpc_server.py
"""
Simple standalone gRPC server for testing.

This script implements a simple gRPC server that can be used to test
the gRPC client in the Django backend. It implements the same interface
as the Go gRPC server but in Python for easier testing.
"""

import os
import sys
import time
import uuid
import logging
import grpc
import signal
from concurrent import futures
from typing import Dict, List, Any, Optional

# Add the protos directory to the Python path
sys.path.append(os.path.join(os.path.dirname(__file__), 'protos'))

# Import the generated protobuf code
from protos import agent_pb2
from protos import agent_pb2_grpc

# Configure logging
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s',
)
logger = logging.getLogger(__name__)


class Task:
    """Represents a task being executed by the agent."""
    
    def __init__(self, task_id: str, prompt: str, context: Dict[str, str], tools: List[str]):
        self.id = task_id
        self.status = "running"
        self.prompt = prompt
        self.context = context
        self.tools = tools
        self.result = ""
        self.events = ["Task created"]


class AgentServiceServicer(agent_pb2_grpc.AgentServiceServicer):
    """Implementation of the AgentService gRPC service."""
    
    def __init__(self):
        self.tasks = {}
        self.lock = futures.ThreadPoolExecutor(max_workers=1)
    
    def ExecuteTask(self, request, context):
        """Execute a task using the agent runtime."""
        logger.info(f"Received ExecuteTask request with prompt: {request.prompt}")
        
        task_id = str(uuid.uuid4())
        
        task = Task(
            task_id=task_id,
            prompt=request.prompt,
            context=request.context,
            tools=request.tools
        )
        
        self.tasks[task_id] = task
        
        # Execute the task asynchronously
        self.lock.submit(self._execute_task_async, task)
        
        return agent_pb2.ExecuteTaskResponse(
            task_id=task_id,
            status="accepted",
            message="Task submitted for execution"
        )
    
    def GetTaskStatus(self, request, context):
        """Get the status of a task."""
        logger.info(f"Received GetTaskStatus request for task: {request.task_id}")
        
        task_id = request.task_id
        
        if task_id not in self.tasks:
            context.set_code(grpc.StatusCode.NOT_FOUND)
            context.set_details(f"Task {task_id} not found")
            return agent_pb2.GetTaskStatusResponse()
        
        task = self.tasks[task_id]
        
        return agent_pb2.GetTaskStatusResponse(
            task_id=task.id,
            status=task.status,
            result=task.result,
            events=task.events
        )
    
    def CancelTask(self, request, context):
        """Cancel a running task."""
        logger.info(f"Received CancelTask request for task: {request.task_id}")
        
        task_id = request.task_id
        
        if task_id not in self.tasks:
            context.set_code(grpc.StatusCode.NOT_FOUND)
            context.set_details(f"Task {task_id} not found")
            return agent_pb2.CancelTaskResponse()
        
        task = self.tasks[task_id]
        
        if task.status == "running":
            task.status = "cancelled"
            task.events.append("Task cancelled")
            
            return agent_pb2.CancelTaskResponse(
                task_id=task.id,
                status="cancelled",
                message="Task cancelled successfully"
            )
        
        return agent_pb2.CancelTaskResponse(
            task_id=task.id,
            status=task.status,
            message=f"Cannot cancel task with status: {task.status}"
        )
    
    def _execute_task_async(self, task):
        """Execute a task asynchronously."""
        time.sleep(2)  # Simulate task execution
        
        if task.status == "cancelled":
            return
        
        task.status = "completed"
        task.result = f"Task completed successfully: {task.prompt}"
        task.events.extend(["Task execution started", "Task execution completed"])


def serve():
    """Start the gRPC server."""
    port = int(os.environ.get('GRPC_SERVER_PORT', '50051'))
    server = grpc.server(futures.ThreadPoolExecutor(max_workers=10))
    
    agent_service = AgentServiceServicer()
    agent_pb2_grpc.add_AgentServiceServicer_to_server(agent_service, server)
    
    server.add_insecure_port(f'[::]:{port}')
    server.start()
    
    logger.info(f"gRPC server started on port {port}")
    
    # Handle graceful shutdown
    def handle_signal(signum, frame):
        logger.info("Received shutdown signal, stopping server...")
        server.stop(0)
    
    signal.signal(signal.SIGINT, handle_signal)
    signal.signal(signal.SIGTERM, handle_signal)
    
    try:
        while True:
            time.sleep(86400)  # Sleep for a day
    except KeyboardInterrupt:
        server.stop(0)


if __name__ == "__main__":
    serve()

File: scripts/simple_grpc_server.go
package main

import (
	"context"
	"fmt"
	"log"
	"net"
	"os"
	"os/signal"
	"sync"
	"syscall"
	"time"

	"github.com/google/uuid"
	"google.golang.org/grpc"
	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/status"
)

type Task struct {
	ID      string
	Status  string
	Prompt  string
	Context map[string]string
	Tools   []string
	Result  string
	Events  []string
}

type AgentServiceServer interface {
	ExecuteTask(context.Context, *ExecuteTaskRequest) (*ExecuteTaskResponse, error)
	GetTaskStatus(context.Context, *GetTaskStatusRequest) (*GetTaskStatusResponse, error)
	CancelTask(context.Context, *CancelTaskRequest) (*CancelTaskResponse, error)
}

type ExecuteTaskRequest struct {
	Prompt  string
	Context map[string]string
	Tools   []string
}

type ExecuteTaskResponse struct {
	TaskId  string
	Status  string
	Message string
}

type GetTaskStatusRequest struct {
	TaskId string
}

type GetTaskStatusResponse struct {
	TaskId  string
	Status  string
	Result  string
	Events  []string
}

type CancelTaskRequest struct {
	TaskId string
}

type CancelTaskResponse struct {
	TaskId  string
	Status  string
	Message string
}

type GRPCServer struct {
	tasks   map[string]*Task
	tasksMu sync.RWMutex
}

func (s *GRPCServer) ExecuteTask(ctx context.Context, req *ExecuteTaskRequest) (*ExecuteTaskResponse, error) {
	log.Printf("Received ExecuteTask request with prompt: %s", req.Prompt)

	taskID := uuid.New().String()

	task := &Task{
		ID:      taskID,
		Status:  "running",
		Prompt:  req.Prompt,
		Context: req.Context,
		Tools:   req.Tools,
		Events:  []string{"Task created"},
	}

	s.tasksMu.Lock()
	s.tasks[taskID] = task
	s.tasksMu.Unlock()

	go s.executeTaskAsync(task)

	return &ExecuteTaskResponse{
		TaskId:  taskID,
		Status:  "accepted",
		Message: "Task submitted for execution",
	}, nil
}

func (s *GRPCServer) GetTaskStatus(ctx context.Context, req *GetTaskStatusRequest) (*GetTaskStatusResponse, error) {
	log.Printf("Received GetTaskStatus request for task: %s", req.TaskId)

	s.tasksMu.RLock()
	task, exists := s.tasks[req.TaskId]
	s.tasksMu.RUnlock()

	if !exists {
		return nil, status.Errorf(codes.NotFound, "task %s not found", req.TaskId)
	}

	return &GetTaskStatusResponse{
		TaskId:  task.ID,
		Status:  task.Status,
		Result:  task.Result,
		Events:  task.Events,
	}, nil
}

func (s *GRPCServer) CancelTask(ctx context.Context, req *CancelTaskRequest) (*CancelTaskResponse, error) {
	log.Printf("Received CancelTask request for task: %s", req.TaskId)

	s.tasksMu.Lock()
	defer s.tasksMu.Unlock()

	task, exists := s.tasks[req.TaskId]
	if !exists {
		return nil, status.Errorf(codes.NotFound, "task %s not found", req.TaskId)
	}

	if task.Status == "running" {
		task.Status = "cancelled"
		task.Events = append(task.Events, "Task cancelled")

		return &CancelTaskResponse{
			TaskId:  task.ID,
			Status:  "cancelled",
			Message: "Task cancelled successfully",
		}, nil
	}

	return &CancelTaskResponse{
		TaskId:  task.ID,
		Status:  task.Status,
		Message: "Cannot cancel task with status: " + task.Status,
	}, nil
}

func (s *GRPCServer) executeTaskAsync(task *Task) {
	time.Sleep(2 * time.Second)

	s.tasksMu.Lock()
	defer s.tasksMu.Unlock()

	if task.Status == "cancelled" {
		return
	}

	task.Status = "completed"
	task.Result = "Task completed successfully: " + task.Prompt
	task.Events = append(task.Events, "Task execution started", "Task execution completed")
}

func RegisterAgentServiceServer(s *grpc.Server, srv AgentServiceServer) {
	s.RegisterService(&_AgentService_serviceDesc, srv)
}

var _AgentService_serviceDesc = grpc.ServiceDesc{
	ServiceName: "agent.AgentService",
	HandlerType: (*AgentServiceServer)(nil),
	Methods: []grpc.MethodDesc{
		{
			MethodName: "ExecuteTask",
			Handler:    _AgentService_ExecuteTask_Handler,
		},
		{
			MethodName: "GetTaskStatus",
			Handler:    _AgentService_GetTaskStatus_Handler,
		},
		{
			MethodName: "CancelTask",
			Handler:    _AgentService_CancelTask_Handler,
		},
	},
	Streams:  []grpc.StreamDesc{},
	Metadata: "agent.proto",
}

func _AgentService_ExecuteTask_Handler(srv interface{}, ctx context.Context, dec func(interface{}) error, interceptor grpc.UnaryServerInterceptor) (interface{}, error) {
	in := new(ExecuteTaskRequest)
	if err := dec(in); err != nil {
		return nil, err
	}
	if interceptor == nil {
		return srv.(AgentServiceServer).ExecuteTask(ctx, in)
	}
	info := &grpc.UnaryServerInfo{
		Server:     srv,
		FullMethod: "/agent.AgentService/ExecuteTask",
	}
	handler := func(ctx context.Context, req interface{}) (interface{}, error) {
		return srv.(AgentServiceServer).ExecuteTask(ctx, req.(*ExecuteTaskRequest))
	}
	return interceptor(ctx, in, info, handler)
}

func _AgentService_GetTaskStatus_Handler(srv interface{}, ctx context.Context, dec func(interface{}) error, interceptor grpc.UnaryServerInterceptor) (interface{}, error) {
	in := new(GetTaskStatusRequest)
	if err := dec(in); err != nil {
		return nil, err
	}
	if interceptor == nil {
		return srv.(AgentServiceServer).GetTaskStatus(ctx, in)
	}
	info := &grpc.UnaryServerInfo{
		Server:     srv,
		FullMethod: "/agent.AgentService/GetTaskStatus",
	}
	handler := func(ctx context.Context, req interface{}) (interface{}, error) {
		return srv.(AgentServiceServer).GetTaskStatus(ctx, req.(*GetTaskStatusRequest))
	}
	return interceptor(ctx, in, info, handler)
}

func _AgentService_CancelTask_Handler(srv interface{}, ctx context.Context, dec func(interface{}) error, interceptor grpc.UnaryServerInterceptor) (interface{}, error) {
	in := new(CancelTaskRequest)
	if err := dec(in); err != nil {
		return nil, err
	}
	if interceptor == nil {
		return srv.(AgentServiceServer).CancelTask(ctx, in)
	}
	info := &grpc.UnaryServerInfo{
		Server:     srv,
		FullMethod: "/agent.AgentService/CancelTask",
	}
	handler := func(ctx context.Context, req interface{}) (interface{}, error) {
		return srv.(AgentServiceServer).CancelTask(ctx, req.(*CancelTaskRequest))
	}
	return interceptor(ctx, in, info, handler)
}

func main() {
	server := grpc.NewServer()

	grpcServer := &GRPCServer{
		tasks: make(map[string]*Task),
	}
	RegisterAgentServiceServer(server, grpcServer)

	port := 50051
	lis, err := net.Listen("tcp", fmt.Sprintf(":%d", port))
	if err != nil {
		log.Fatalf("Failed to listen on port %d: %v", port, err)
	}

	sigCh := make(chan os.Signal, 1)
	signal.Notify(sigCh, os.Interrupt, syscall.SIGTERM)
	go func() {
		<-sigCh
		log.Println("Received shutdown signal, stopping server...")
		server.GracefulStop()
	}()

	log.Printf("Starting gRPC server on port %d...", port)
	if err := server.Serve(lis); err != nil {
		log.Fatalf("Failed to serve: %v", err)
	}
}
